core/src/main/java/com/oath/oak/NativeAllocator/OakNativeMemoryAllocator.java

```diff
@@ -57,83 +71,83 @@
     // constructor
     // input param: memory capacity given to this Oak. Uses default BlocksPool
     public OakNativeMemoryAllocator(long capacity) {
-      this(capacity, BlocksPool.getInstance());
+        this(capacity, BlocksPool.getInstance());
     }
 
     // A testable constructor
     OakNativeMemoryAllocator(long capacity, BlocksProvider blocksProvider) {
         this.blocksProvider = blocksProvider;
-        int blockArraySize = ((int)(capacity/blocksProvider.blockSize()))+1;
+        int blockArraySize = ((int) (capacity / blocksProvider.blockSize())) + 1;
         // first entry of blocksArray is always empty
-        this.blocksArray = new Block[blockArraySize+1];
+        this.blocksArray = new Block[blockArraySize + 1];
         // initially allocate one single block from pool
         // this may lazy initialize the pool and take time if this is the first call for the pool
         allocateNewCurrentBlock();
         this.capacity = capacity;
+        for (int i = 0; i < ThreadIndexCalculator.MAX_THREADS; i++) {
+            dummies[i] = new FreeChuck(-1, 0, null);
+        }
     }
 
     @Override
     public ByteBuffer allocate(int size) {
-      return allocate(size, true);
+        return allocateSlice(size).getByteBuffer();
     }
 
     // Allocates ByteBuffer of the given size, either from freeList or (if it is still possible)
     // within current block bounds.
     // Otherwise new block is allocated within Oak memory bounds. Thread safe.
-    private ByteBuffer allocate(int size, boolean reuseFreedMemory) {
-
-      if (reuseFreedMemory) {
-        if (!freeList.isEmpty()) {
-          for (Map.Entry<Long, Slice> kv : freeList) {
-            ByteBuffer bb = kv.getValue().getByteBuffer();
-            if (bb.remaining() > (RECLAIM_FACTOR * size)) break;     // all remaining buffers are too big
-            if (bb.remaining() >= size && freeList.remove(kv)) {
-              if (stats != null) stats.reclaim(size);
-              return bb;
+    @Override
+    public Slice allocateSlice(int size) {
+
+        FreeChuck myDummy = dummies[threadIndexCalculator.getIndex()];
+        while (!freeList.isEmpty()) {
+            myDummy.length = size;
+            FreeChuck bestFit = freeList.higher(myDummy);
+            if (bestFit == null) {
+                break;
+            }
+            if (bestFit.slice.getByteBuffer().remaining() > (RECLAIM_FACTOR * size)) {
+                break;     // all remaining buffers are too big
+            }
+            if (freeList.remove(bestFit)) {
```

Comment:
freeList is a concurrent structure, is it that many threads can get the same bestFit chunk as an output of line 106? Is it that only one will succeed to remove the bestFit and get the allocation? Others will allocate a new space, although they might have yet another allocation from freeList, am I wrong?

Reply:
Thread re-execute this loop. I added this explanation in a comment.